Sources/NIOHTTP1/NIOTypedHTTPServerUpgraderStateMachine.swift

```diff
@@ -0,0 +1,360 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import DequeModule
+import NIOCore
+
+@available(macOS 13, iOS 16, tvOS 16, watchOS 9, *)
+struct NIOTypedHTTPServerUpgraderStateMachine<UpgradeResult> {
+    enum State {
+        /// The state before we received a TLSUserEvent. We are just forwarding any read at this point.
+        case initial
+
+        /// The request head has been received. We're currently running the future chain awaiting an upgrader.
+        case awaitingUpgrader(
+            seenFirstRequest: Bool,
+            buffer: Deque<NIOAny>
+        )
+
+        /// We have an upgrader, which means we can begin upgrade we are just waiting for the request end.
+        case upgraderReady(
+            upgrader: any NIOTypedHTTPServerProtocolUpgrader<UpgradeResult>,
+            requestHead: HTTPRequestHead,
+            responseHeaders: HTTPHeaders,
+            proto: String,
+            buffer: Deque<NIOAny>
+        )
+
+        /// We are either running the upgrading handler.
+        case upgrading(
+            buffer: Deque<NIOAny>
+        )
+
+        case unbuffering(buffer: Deque<NIOAny>)
+
+        case finished
+
+        case modifying
+    }
+
+    private var state = State.initial
+
+    @usableFromInline
+    enum HandlerRemovedAction {
+        case failUpgradePromise
+    }
+
+    @inlinable
```

Comment:
We're very inconsistent with the `@inlinable`s here, despite all the actions being `@usableFromInline`. Why is that?